### DISCUSSION TITLE
Pydantic, Annotated, and v1.1.348

### DISCUSSION THREAD

**Discussion by amacfie-tc** [microsoft/pyright#7091] [Q&A]
Prior to v1.1.348, the following had no type errors (and worked)
```python
# Pydantic 2
from pydantic import AnyHttpUrl, TypeAdapter

url = TypeAdapter(AnyHttpUrl).validate_python("http://example.com")
```
Now it gives `No overloads for "__new__" match the provided arguments (reportCallIssue)`. The code for `TypeAdapter` is [here](https://github.com/pydantic/pydantic/blob/b2aa36a80904773d7db2e4c007fc00a5a4d7653b/pydantic/type_adapter.py#L111) and `AnyHttpUrl` is defined as `Annotated[Url, UrlConstraints(allowed_schemes=['http', 'https'])]` [here](https://github.com/pydantic/pydantic/blob/b2aa36a80904773d7db2e4c007fc00a5a4d7653b/pydantic/networks.py#L103).

The Pydantic docs [here](https://docs.pydantic.dev/2.5/concepts/types/#composing-types-via-annotated) contain a code example that now has the above type error.

Should this indeed be a type error or is Pydantic abusing the type system?

**Comment by erictraut** [comment]
Hmm, I think this is an ambiguous part of the typing spec.

You are attempting to pass an `Annotated` class to a parameter typed as `type[T]`. It's not clear whether this should be permitted. `Annotated` is a special form, and most special forms are not type compatible with `type[T]`. I recently tightened up pyright's enforcement of this, which is why you're now seeing this error. It appears that mypy does allow `Annotated`, and it also appears that the `Annotated` class implementation forwards some calls to its first type argument, but it doesn't forward all calls. That means an `Annotated` class can't be used, for example, as the second argument to `isinstance`.

I'll start a discussion in the Python typing forums to get feedback from the broader typing community. Once we come to a consensus on this issue, I'll try to revise the official typing spec and (if appropriate) update pyright accordingly.

**Reply by erictraut** [reply]
For reference, here's a link to the [discussion thread](https://discuss.python.org/t/is-annotated-compatible-with-type-t/43898).

**Reply by Viicos** [reply]
For reference, I recently updated Pydantic's `TypeAdapter` signature, but this was prior to v1.1.346 which introduced the change.

Here is the PR: https://github.com/pydantic/pydantic/pull/8453. In particular, these two comments: https://github.com/pydantic/pydantic/pull/8453#issuecomment-1886757491 / https://github.com/pydantic/pydantic/pull/8453#issuecomment-1889094862

cc @adriangb

**Reply by samuelcolvin** [reply]
As per #7228 this is a problem without `Annotated` - in that case a union is causing the problem.

**Reply by Viicos** [reply]
Sorry my point was that both `Annotated` and `Union` are special forms, and pyright recently introduced some changes regarding special forms.

But it might be that `Annotated` should be considered differently than `Union`? I believe in the current context (that is, using it as a value instead of an annotation) they are just considered as special forms and no special casing is applied to one or the other
